src/constants/label.js

```diff
@@ -117,13 +117,134 @@ export function localizeLayers(layers, locales) {
 }
 
 /**
- * The name in the user's preferred language.
+ * Returns an expression that replaces a finite number of occurrences of a
+ * substring expression withing a larger string expression, starting at a given
+ * index.
+ *
+ * This expression nests recursively by the maximum number of replacements. Take
+ * special care to minimize this limit, which exponentially increases the length
+ * of a property value in JSON. Excessive nesting causes acute performance
+ * problems when loading the style.
+ *
+ * The returned expression can be complex, so use it only once within a property
+ * value. To reuse the evaluated value, bind it to a variable in a let
+ * expression.
+ *
+ * @param haystack The overall string expression to search within.
+ * @param needle The string to search for, or an expression that evaluates to
+ *  this string.
+ */
+export function replaceExpression(
+  haystack,
+  needle,
+  replacement,
+  haystackStart,
+  numReplacements = 1
+) {
+  let asIs = ["slice", haystack, haystackStart];
+  if (numReplacements <= 0) {
+    return asIs;
+  }
+
+  let needleStart = ["index-of", needle, haystack, haystackStart];
+  let needleLength =
+    typeof needle === "object" ? ["length", needle] : needle.length;
+  let needleEnd = ["+", needleStart, needleLength];
+  return [
+    "case",
+    [">=", needleStart, 0],
+    [
+      "concat",
+      ["slice", haystack, haystackStart, needleStart],
+      replacement,
+      replaceExpression(
+        haystack,
+        needle,
+        replacement,
+        needleEnd,
+        numReplacements - 1
+      ),
+    ],
+    asIs,
+  ];
+}
+
+/**
+ * Maximum number of values in a semicolon-delimited list of values.
+ *
+ * Increasing this constant deepens recursion for replacing delimiters in the
+ * list, potentially affecting style loading performance.
+ */
+const maxValueListLength = 9;
+
+/**
+ * Returns an expression interpreting the given string as a list of tag values,
+ * pretty-printing the standard semicolon delimiter with the given separator.
+ *
+ * https://wiki.openstreetmap.org/wiki/Semi-colon_value_separator
+ *
+ * The returned expression can be complex, so use it only once within a property
+ * value. To reuse the evaluated value, bind it to a variable in a let
+ * expression.
+ *
+ * @param valueList A semicolon-delimited list of values.
+ * @param separator A string to insert between each value, or an expression that
+ *  evaluates to this string.
+ */
+export function listValuesExpression(valueList, separator) {
+  let maxSeparators = maxValueListLength - 1;
+  // Replace the ;; escape sequence with a placeholder sequence unlikely to
+  // legitimately occur inside a value or separator.
+  const objReplacementChar = "\x91\ufffc\x92"; // https://overpass-turbo.eu/s/1pJx
```

Comment:
It’s amazing how much junk there is in `name` that can impersonate a [strip marker](https://www.mediawiki.org/wiki/Strip_marker), but this sequence does the trick for now. If someone opens a pub for text processing geeks and names it “The &#x91;&#xfffc;&#x92; & ;”, I’ll kindly ask someone to offer up a prized password as sacrifice.